src/Streamly/Internal/Unicode/Stream.hs

```diff
@@ -423,6 +427,80 @@ data CodingFailureMode
 replacementChar :: Char
 replacementChar = '\xFFFD'
 
+data UTF8CharDecodeState a
+    = UTF8CharDecodeInit
+    | UTF8CharDecoding !DecodeState !CodePoint
```

Comment:
If it does not fuse you can try ANN Fuse on it.

Reply:
It fuses without the `ANN`